pandas/core/groupby/ops.py

```diff
@@ -637,43 +671,25 @@ def _cython_operation(
             if not is_complex_dtype(dtype):
                 values = ensure_float64(values)
 
-        arity = self._cython_arity.get(how, 1)
         ngroups = self.ngroups
+        comp_ids, _, _ = self.group_info
 
         assert axis == 1
         values = values.T
-        if how == "ohlc":
-            out_shape = (ngroups, 4)
-        elif arity > 1:
-            raise NotImplementedError(
-                "arity of more than 1 is not supported for the 'how' argument"
-            )
-        elif kind == "transform":
-            out_shape = values.shape
-        else:
-            out_shape = (ngroups,) + values.shape[1:]
-
-        func, values = self._get_cython_func_and_vals(kind, how, values, is_numeric)
-
-        if how == "rank":
-            out_dtype = "float"
-        else:
-            if is_numeric:
-                out_dtype = f"{values.dtype.kind}{values.dtype.itemsize}"
-            else:
-                out_dtype = "object"
 
-        codes, _, _ = self.group_info
+        out_shape = WrappedCythonOp.get_output_shape(how, kind, ngroups, values)
```

Comment:
i would go one step further and use an instance of WrappedCythonOp here, then you can pass whatever you need into the construct and just ask the instance for properties (if you want to make this simple), you can use a dataclass for example.

I find this pattern *better* than using a class like this (meaning the class is just a collection of methods here).

Reply:
updated + green